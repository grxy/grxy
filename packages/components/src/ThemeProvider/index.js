import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming'
import { node } from 'prop-types'
import React from 'react'

import { getTheme } from '@grxy/theme'

const theme = getTheme()

const ThemeProvider = ({ children }) => (
    <EmotionThemeProvider theme={theme}>{children}</EmotionThemeProvider>
)

ThemeProvider.propTypes = {
    children: node.isRequired,
}

export default ThemeProvider
