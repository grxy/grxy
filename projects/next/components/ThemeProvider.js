import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming'
import { node, string } from 'prop-types'
import React, { Component } from 'react'

import { getTheme } from '../lib/theme'
class ThemeProvider extends Component {
    static defaultProps = {
        serverTime: undefined,
    }

    static propTypes = {
        children: node.isRequired,
        serverTime: string,
    }

    state = getTheme(new Date(this.props.serverTime))

    componentDidMount() {
        /*
            The intent of the following code is to refresh the theme at the top
            of every hour. I'm sure there are simpler ways to do this using
            moment or date-fns, but I don't want to add a date lib just yet.

            TODO: convert to hook: useScheduledCallback or useHourlyCallback
        */
        const now = new Date()

        const next = new Date(now)
        next.setHours(now.getHours() + 1)
        next.setMinutes(0)
        next.setSeconds(0)
        next.setMilliseconds(0)

        const timeout = next - now

        this.refreshTheme()

        this.timeout = setTimeout(() => {
            this.interval = setInterval(() => {
                this.refreshTheme()
            }, 36e5) // every 1 hour

            this.refreshTheme()
        }, timeout)
    }

    componentWillUnmount() {
        clearTimeout(this.timeout)
        clearInterval(this.interval)
    }

    refreshTheme() {
        this.setState(() => getTheme())
    }

    render() {
        return (
            <EmotionThemeProvider theme={this.state}>
                {this.props.children}
            </EmotionThemeProvider>
        )
    }
}

export default ThemeProvider
