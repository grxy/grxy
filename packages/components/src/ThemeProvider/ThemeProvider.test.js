import { mount } from 'enzyme'
import React from 'react'

import ThemeProvider from './'

jest.mock('@grxy/theme', () => ({
    getTheme: () => ({
        test: true,
    }),
}))

describe('<ThemeProvider />', () => {
    let wrapper
    beforeEach(() => {
        wrapper = mount(
            <ThemeProvider>
                <></>
            </ThemeProvider>,
        )
    })

    it('renders a theme based on the current time', () => {
        expect(wrapper).toMatchSnapshot()
    })
})
