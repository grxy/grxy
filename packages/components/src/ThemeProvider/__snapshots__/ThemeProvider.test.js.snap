// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeProvider /> renders a theme based on the current time 1`] = `
<ThemeProvider>
  <ThemeProvider
    theme={
      Object {
        "test": true,
      }
    }
  />
</ThemeProvider>
`;
